src/libraries/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.SecKeyRef.macOS.cs

```diff
@@ -36,7 +36,7 @@ internal static partial class AppleCrypto
 
         internal static SafeSecKeyRefHandle ImportEphemeralKey(ReadOnlySpan<byte> keyBlob, bool hasPrivateKey)
         {
-            Debug.Assert(keyBlob != null);
+            Debug.Assert(keyBlob != default);
```

Comment:
I think it would be better to assert that the spans are not empty, and then reevaluate whether asserting it makes sense for each case.

Reply:
CC @bartonjs @vcsjones

Reply:
Comparing Span with `default` is equally problematic as comparing with `null`. The analyzer should flag comparing with `default` as well.

Reply:
We should be careful about "null" / "default" spans vs. empty. There are several places in S.S.Cryptography where a zero length span is handled differently depending on if the ref of the span is null or not.

For example, in some places, an empty span is "okay". A "null" span is not. This is because when a "null" span is pinned, the pinned pointer is nullptr, which is what we may be trying to assert against.

"default" is fine, I think, but less clear what the intention is.

Reply:
It would be nice to have an explicitly comment on the few places where this `null` or `default` is used intentionally in the comparison.

Reply:
> Comparing Span with `default` is equally problematic as comparing with `null`. The analyzer should flag comparing with `default` as well.

Now we are detecting `null` and suggesting changing it with `default` or use `IsEmpty`, apparently we should not suggest `default` too, another suggestion (code-fix) could be removing the condition, is there anything else we could suggest?

Reply:
> For example, in some places, an empty span is "okay". A "null" span is not. This is because when a "null" span is pinned, the pinned pointer is nullptr, which is what we may be trying to assert against.

@vcsjones could you make those suggestions on each security related updates, or should I exclude the security related the changes from this PR

Reply:
> I think such cases should check the ref with `Unsafe.IsNullRef` explicitly instead of checking `default`.

Well  `Unsafe.IsNullRef<T>(ref T)` requires type parameter, and I cannot use `ReadOnlySpan<T>` as type parameter.

Reply:
Ah, looks I supposed to use `MemoryMarshal.GetReference<T>`

Reply:
yeah. Though, I would not be surprised if some of this null-checking was in packages that target .NET Standard, where those APIs are not available.

Reply:
Doesn't seem build on .NET Standard, updated, will see if fails on any CI legs

Reply:
> Comparing Span with default is equally problematic as comparing with null. The analyzer should flag comparing with default as well.

For what it's worth, during the review session where we approved the analyzer we explicitly said that we wanted to block `== null`, but leave `== default` alone.  We probably shouldn't have changed the scope of the analyzer without further group review; but :shrug:.

https://github.com/dotnet/runtime/issues/84265#issuecomment-1650277489

Reply:
Ah ok ... I did not know that this point was discussed before.

I believe that 99+% of code out there should only ever check for empty spans. The distinction between empty default span and empty non-default span is very subtle, and it tends to only matter for specific low-level code. We do not generally specify when our public APIs return empty non-default span vs. empty default span. 

If we wanted to say that comparing Span with `default` is fine in ordinary code, we may want to start explicitly thinking about when our APIs return empty non-default span vs. empty default span.

Reply:
> For what it's worth, during the review session where we approved the analyzer we explicitly said that we wanted to block == null, but leave == default alone. We probably shouldn't have changed the scope of the analyzer without further group review; but 🤷.

In general, we do follow the approved behavior, we test the analyzer implementation with runtime or other repos to fine tune the analyzer, this change felt like an improvement and @stephentoub suggested the same (offline). Did not really think about to bring back for review. I can bring it back and update the analyzer accordingly.